---
title: "R Code for Data analysis"
author: "Juliette Bellengier, Olivier Tenaillon"
date: "11/01/2025"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

### **High-Throughput Conjugation Reveals Strain Specific Recombination Patterns Enabling Precise Trait Mapping in Escherichia coli**

------------------------------------------------------------------------

###### [Last revised by:]{.underline} Juliette Bellengier 16/01/2024

------------------------------------------------------------------------

[Packages & Libraries installation]{.underline}

```{r setup, include=FALSE}
# Définir un miroir CRAN
options(repos = c(CRAN = "https://cran.rstudio.com"))
install.packages("ggplot2")
install.packages("scale")
install.packages("HMM")
install.packages("ggpubr")
install.packages("viridis")
install.packages("viridisLite")
install.packages("dplyr")
install.packages("ggpmisc")
```

```{r}
library(HMM)
library(dplyr)
library(ggpmisc)
require(lattice)
library(viridisLite)
library(viridis)
library(ggplot2)
library(ggpubr)
library(readxl)
library(dplyr)
library(RColorBrewer)
library(scales)
```

------------------------------------------------------------------------

[Figure 2.B]{.underline}

```{r}
F2<- read_excel("Fig1a3_paper/F2_paper.xlsx")
F2$CFUdonnor <- log10(as.numeric(F2$CFUdonnor))
F2_guide1 = F2[F2$guide =='sacB', ] 
F2_guide1 <- F2_guide1 %>%
  mutate(across(1, ~ recode(., "A" = " BW25113"))) 

tiff("PhD/Figures/Figure2.tiff", width = 4.5, height = 4.5, units = "in", res = 500)


ggplot(F2_guide1, aes(x = genotype, y = CFUdonnor, fill = guide)) + 
  labs(y = "log(CFU Hfr/pNMT3 donor)") +
  geom_boxplot(color = "black", size = 0.5) +  
  geom_point(position = position_dodge(width = 0.8),  
             shape = 21, alpha = 0.6, stroke = 1, fill = "#b1e9e3", color = "black") + 
  scale_fill_manual(values = c("#b1e9e3")) +  
  theme_bw() +  # Fond blanc
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    legend.position = "none", 
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),  
    panel.border = element_blank()  
  ) +
  geom_vline(xintercept = seq(0.5, length(unique(F2_guide1$genotype)) - 0.5, by = 1), 
             color = "gray", size = 0.3, linetype = "solid")
```

------------------------------------------------------------------------

[Supplementary Figure 1]{.underline}

```{r}
F3B <- read_excel("Fig1a3_paper/F3B.xlsx")
F3B$CFUdonnor <- log10(as.numeric(F3B$CFUdonnor))
F3B <- F3B %>% mutate(system = recode(system, "psim5" = "control"))
F3B$CFUdonnor <- replace(F3B$CFUdonnor, F3B$CFUdonnor == -Inf, -10)
F3B$system <- factor(F3B$system, levels = c("control", "cas9"))

tiff("PhD/Figures/Suppplement1.tiff", width = 4.5, height = 4.5, units = "in", res = 500)
ggplot(F3B, aes(x = genotype, y = CFUdonnor, fill = guide)) + 
  labs(y = "log(CFU Hfr/pNMT3 donor)") +
  geom_boxplot(aes(fill = guide), color = "black", linewidth = 0.5, position = position_dodge(width = 0.8)) + 
  geom_point(aes(fill = guide),  
             position = position_dodge(width = 0.8),  
             shape = 21, alpha = 0.6, stroke = 1, color = "black") +
  scale_fill_manual(values = c("sacB" = "orange", "sgrna" = "#ffd846"))+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  theme_bw() +  
  theme(
    axis.ticks.x = element_line(color = "black", size = 0.5),
    legend.position = "none"
  ) +
  geom_vline(xintercept = seq(0.5, length(unique(F3B$genotype)) - 0.5, by = 1), 
             color = "black", size = 0.3, linetype = "solid")

```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

[Figure 3.B]{.underline}

```{r}
F2 <- F2 %>%
  mutate(across(1, ~ recode(., "A" = " BW25113"))) 
F2$guide <- factor(F2$guide)  


###test normalité
F2_testnorm <- F2 %>%
  filter(genotype == "rfe")
shapiro_results <- shapiro.test(F2_testnorm$CFUdonnor)
print(shapiro_results) # pvalue < 0.05 data non normales donc pas KW test
######

tiff("PhD/Figures/Figure3C.tiff", width = 4.5, height = 4.5, units = "in", res = 500)
ggplot(F2, aes(x = genotype, y = CFUdonnor, fill = guide)) + 
  labs(y = "log(CFU Hfr/pNMT3 donor)") +
  geom_boxplot(color = "black", size = 0.5, position = position_dodge(width = 0.8)) + 
  geom_point(aes(fill = guide),  
             position = position_dodge(width = 0.8),  
             shape = 21, alpha = 0.8, stroke = 1, color ="black") + 
  scale_fill_manual(values = c("#b1e9e3", "#cbb1e9")) +  
  scale_color_manual(values = c("#b1e9e3", "#cbb1e9")) + 
  stat_compare_means(aes(group = guide), label = "p.signif", label.y = -1, 
                     method = "wilcox.test") +  
  theme_bw() +  
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),  
    legend.position = "none",  
    panel.grid.major = element_blank(),  
    panel.grid.minor = element_blank(),  
    panel.border = element_blank()  
  ) +
  geom_vline(xintercept = seq(0.5, length(unique(F2$genotype)) - 0.5, by = 1), 
             color = "gray", size = 0.3, linetype = "solid")
```

------------------------------------------------------------------------

[Figure 4.C]{.underline}

```{r}
strains <- c("K12", "K12 Library", "REL606", "REL606 Library")
replicate1 <- c(4720000, 59000, 27800000, 83000)
replicate2 <- c(2070000, 24000, 40400000, 125000)

F4B <- data.frame(Strain = strains, 
                   Replicate1 = replicate1, 
                   Replicate2 = replicate2)

F4B <- F4B %>%
  gather(key = "Replicate", value = "Value", -Strain)
F4B_mean <- F4B %>%
  group_by(Strain) %>%
  summarise(moyenne = mean(Value, na.rm = TRUE))

ggplot(F4B, aes(x = Strain, y = log10(Value))) +
  labs(y = "Log (Recombination CFU/mL)") +  # Modifier le label de l'axe Y
  geom_point(color = "gold2", size = 3) +  # Points pour les réplicats
  geom_point(data = F4B_mean, aes(x = Strain, y = log10(moyenne)), shape = 17, color = "darkblue", size = 4) +  
  scale_y_continuous(trans = "log", labels = scales::math_format(e^.x)) +  # Transformation log en base e
  theme_light() + 
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    legend.position = "none"
  )+ 
  stat_compare_means(aes(group = Strain), label = "p.signif", label.y = -1, 
                     method = "wilcox.test") 

```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

[Figure 4.B]{.underline}

```{r}
F5B<- read_excel("Fig1a3_paper/F5B.xlsx")
F5B$distance <- as.factor(F5B$distance)
F5B$transconjugant <- as.numeric(F5B$transconjugant)

tiff("PhD/Figures/Figure4C.tiff", width = 4.5, height = 4.5, units = "in", res = 500)


ggplot(F5B, aes(x = distance, y = transconjugant)) + 
  labs(y = "log10(CFU galK+ \ntransconjugants/galk- recipient)", 
       x = "Physical distance (in kb)") + 
  geom_boxplot(fill = "#FFB74D") +  
  geom_point(fill="#FFB74D", color = "black",  
             shape = 21,  
             alpha = 0.7,  
             stroke = 1,  
             position = position_jitter(width = 0.2)) +  
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  theme_bw() +  # Fond blanc
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    legend.position = "none"  
  )+ 
  stat_compare_means(aes(group = distance), 
                     method = "wilcox.test", 
                     comparisons = list(c("100", "200"), c("100", "300"), c("100", "939"),c("100", "1000"),
                                        c("100", "3108"),
                                        c("100", "3177")), 
                     label = "p.signif", tip.length = 0, exact = FALSE)

```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

[Figure 4.D]{.underline}

```{r}
F4D <- read_excel("Fig1a3_paper/F5D.xlsx")
F4D <- subset(F4D, F4D[, 1] == "K12")
colnames(F4D) <- c("Donor", "Recipient", "CFU_rec")
F4D$Recipient <- as.factor(F4D$Recipient)
F4D$CFU_rec <- as.numeric(F4D$CFU_rec)

F4D_mean <- F4D %>%
  group_by(Recipient) %>%
  summarise(moyenne = mean(CFU_rec, na.rm = TRUE))

F4D$Recipient <- reorder(F4D$Recipient, F4D$CFU_rec, FUN = mean)
F4D$Recipient <- factor(F4D$Recipient, levels = rev(levels(F4D$Recipient)))
tiff("PhD/Figures/Figure4D.tiff", width = 4.5, height = 4.5, units = "in", res = 500)

ggplot(F4D, aes(x = Recipient, y = CFU_rec)) +
  labs(x = "Recipient", y = "Log10 (Recombination \n Efficiency)") + 
  geom_point(color = "gold2", size = 3) + 
  geom_point(data = F5D_mean, aes(x = Recipient, y = moyenne), shape = 17, color = "darkblue", size = 4) + 
  scale_y_log10(labels = trans_format("log10", scales::math_format(10^.x))) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")

dev.off()
```

------------------------------------------------------------------------

[Figure 5.B]{.underline}

```{r echo=TRUE, warning=FALSE, message=FALSE}
library(HMM)
extrafigures=0
FragmentPosions<-data.frame(recombinant=c(0),rec=c(0),begin=c(0),end=c(0),nb1=c(0),nbpos1=c(0))
i<-1
k_vec = c(1:6, 8, 8, 9, 11:20)
u_vec = c(1:20)
j_vec = c(9:15, 17, 18, 19:28, 29:56)
w_vec = c(1:8)
for (v in 1:length(j_vec)) {
  j = j_vec[v]
  k=k_vec[v]
  u=u_vec[v]
  print(v)
  if (v >= 40) {
    w = v - 39 
  } else {
    w = 1 
  }
  
  #REL606 recipient
  galkpos=769500
  Donor<-seq(0,5000000,1000)*0
  Recipient<-seq(0,5000000,1000)*0
  #HS Recipient
  if(j > 28){
    galkpos = 816000
    Donor<-seq(0,5000000,1000)*0
    Recipient<-seq(0,5000000,1000)*0
  }
  #536 recipient
  if (j>48){
    galkpos = 803500
    Donor<-seq(0,5500000,1000)*0
    Recipient<-seq(0,5500000,1000)*0
  }
  
  
  Donor<-Donor[-1]
  Recipient<-Recipient[-1]
  
  
  if (j<29){
    name2=paste("position/GalK_Rel/", "K12_sample",j, "relgal", k,"_pos.txt",sep="")
    name1=paste("position/GalK_Rel/","Rel606_sample",j, "relgal", k,"_pos.txt",sep="")}
  else{
    if (j<49){
      if (v > 19){
        k=u_vec[v-19]
        u=u_vec[v-19]
      }
      name2=paste("position/GalK_HS/", "K12_sample",j, "HSgal", u,"_pos.txt",sep="")
      name1=paste("position/GalK_HS/","HS_sample",j, "HSgal", u,"_pos.txt",sep="")}
    else{
      name2=paste("position/GalK_536_new/", "K12_sample",j, "_536gal", w,"_pos.txt",sep="")
      name1=paste("position/GalK_536_new/","536_sample",j, "_536gal", w,"_pos.txt",sep="")
      print(name1)
      print(w)
    }}
  Tab<-read.table(name1)
  Tab2<-read.table(name2)
  
  #exclude problematic regions
  Tab2<-Tab2[Tab2[,2]<3947000 | Tab2[,2]>=3949000,]
  Tab2<-Tab2[Tab2[,2]<3481000| Tab2[,2]>=3482000,]
  Tab<-Tab[Tab[,1]<3947000 | Tab[,1]>=3949000,]
  Tab<-Tab[Tab[,1]<3481000| Tab[,1]>=3482000,]
  
  if (j>48)
  {
    Tab<-Tab[Tab[,1]<=804500 |Tab[,1]>=805000 ,]
    Tab2<-Tab2[Tab2[,2]<=804500 | Tab2[,2]>=805000,]
  }
  #filter the  weird positions: 3947406 and  3481583
  
  if (j>28 & j<49) 
  {
    Tab<-Tab[Tab[,1]<4012000 | Tab[,1]>=4014000,]
    Tab2<-Tab2[Tab2[,2]<4012000| Tab2[,2]>=4014000,]
    Tab<-Tab[Tab[,1]<2315000 | Tab[,1]>=2318000,]
    Tab2<-Tab2[Tab2[,2]<2315000| Tab2[,2]>=2318000,]
  }
  
  # Bin data for figures with error handling for histograms
  a <- hist(Tab[Tab[, 2] >= 1e-1 & Tab[, 2] <= 5e+06, 1], breaks = seq(0, 5000000, 1000), plot = F)
  b <- hist(Tab2[Tab2[, 2] >= 2e-1 & Tab2[, 2] <= 5e+06, 2], breaks = seq(0, 5000000, 1000), plot = F)
  
  # Ajuster les breaks pour couvrir toute la plage des données
  breaks = seq(min(Tab2[, 2]), max(Tab2[, 2]), by = 1000)
  
  Donor <- Donor + b$counts
  Recipient <- Recipient + a$counts
  
  # Combine all reads according to the recipient position and attribute to who they matched, keeping only reads that matched both genomes but preferentially one
  TabSorted1 <- as.data.frame(Tab[Tab[, 2] >= 0 & abs(Tab[, 1] - Tab[, 2]) < 500000, 1]) #cette ligne PROBLEMOOOOOO
  if (nrow(TabSorted1) == 0) {
    print("TabSorted1 est vide, passage à la prochaine itération")
    next  # Passer à l'itération suivante de la boucle
  }
  TabSorted1$source <- 0 # 0 means recipient match
  dimnames(TabSorted1)[[2]][1] <- "position"
  
  TabSorted2 <- as.data.frame(Tab2[Tab2[, 2] >= 0 & abs(Tab2[, 1] - Tab2[, 2]) < 500000, 2])
  TabSorted2$source <- 1   # 1 means donor match
  dimnames(TabSorted2)[[2]][1] <- "position"
  
  # Combine the sorted data from both tables
  TabSortedCompiled <- rbind(TabSorted1, TabSorted2)
  
  # Sorting the combined table by position
  TabSortedCompiled <- TabSortedCompiled[order(TabSortedCompiled$position),]
  
  
  if (j<49) {      
    hmm=initHMM(c("MG","REL","Mix"),c(0,1),startProbs=c(0.1,0.9,0.1),transProbs=matrix(c(0.9999999,0.0000001,0.0000001,0.0000001,0.9999999,0.0000001,0.0000001,0.0000001,0.9999999),3),emissionProbs=matrix(c(0.1,0.9,0.5,0.9,0.1,0.5),3))
  }else
  {
    #  hmm=initHMM(c("MG","REL"),c(0,1),startProbs=c(0.1,0.9),transProbs=matrix(c(0.9999999,0.0000001,0.0000001,0.9999999),2),emissionProbs=matrix(c(0.1,0.9,0.9,0.1),2))
    hmm=initHMM(c("MG","REL","Mix"),c(0,1),startProbs=c(0.1,0.9,0.1),transProbs=matrix(c(0.9999999,0.000000001,0.000000001,0.000000001,0.999999999,0.000000001,0.000000001,0.000000001,0.999999999),3),emissionProbs=matrix(c(0.1,0.9,0.5,0.9,0.1,0.5),3))
  }
  
    #posterior=posterior(hmm,as.factor(TabSorted$source))
  viterbihmm=viterbi(hmm,as.factor(TabSortedCompiled$source))
  TabSortedCompiled$State<-viterbihmm
  TabSortedCompiled$Donor<-as.numeric(TabSortedCompiled$State!="REL")
  
  #plot figure with fit
  #print("figure")
  TabSortedCompiled$cumsum<-cumsum(as.numeric(TabSortedCompiled$source*2-1))
  
  
    if (v == 1) {
      tiff("PhD/Figures/Figure5B.tiff", width = 4.5, height = 4.5, units = "in", 
           res = 500)
      plot(a$mids/1e6,log10(Recipient),pch=20,cex=1,
           xlab="Genome position (Mb)",
           ylab="log10(coverage)",ylim=c(0,max(log10(Recipient),log10(Donor))),
             col="#155289",cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5)
        abline(v=galkpos/1e6,col="red", lwd=3)
      # plot(a$mids/1e6,log10(Recipient),pch=20,cex=1,main=paste("Recombinant",j),
      #      xlab="Genome osition (Mb)",
      #      ylab="log10(coverage)",ylim=c(0,max(log10(Recipient),log10(Donor))),
      #        col="#155289",cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5)
  points(b$mids/1e6,log10(Donor),pch=20,cex=1,col= "#F4E6AA")
  maximim<-max(log10(Donor),log10(Recipient))
  lines(TabSortedCompiled$position/1e6,TabSortedCompiled$Donor*maximim,col="red",lwd=2)
  dev.off()}
  
      if (v == 32) {
      tiff("PhD/Figures/Figure5B_40.tiff", width = 4.5, height = 4.5, units = "in", 
           res = 500)
      plot(a$mids/1e6,log10(Recipient),pch=20,cex=1,
           xlab="Genome position (Mb)",
           ylab="log10(coverage)",ylim=c(0,max(log10(Recipient),log10(Donor))),
             col="#155289",cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5)
        abline(v=galkpos/1e6,col="red", lwd=3)
      # plot(a$mids/1e6,log10(Recipient),pch=20,cex=1,main=paste("Recombinant",j),
      #      xlab="Genome osition (Mb)",
      #      ylab="log10(coverage)",ylim=c(0,max(log10(Recipient),log10(Donor))),
      #        col="#155289",cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5)
  points(b$mids/1e6,log10(Donor),pch=20,cex=1,col= "#F4E6AA")
  maximim<-max(log10(Donor),log10(Recipient))
  lines(TabSortedCompiled$position/1e6,TabSortedCompiled$Donor*maximim,col="red",lwd=2)
  dev.off()
  }
  
  #   if(extrafigures==1)
  # {
  #   #plot zoom  with fit FOR REL606 and HS 
  #   plot(a$mids/1e6,log10(Recipient),pch=20,cex=1,main=paste("Recombinant",j),xlab="genome position (Mb)",ylab="log10(coverage)",ylim=c(0,2),col="#155289",cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5,xlim=c(0.7,1))
  #   # points(b$mids,log10(Donor),pch=20,cex=1,col="blue")
  #   points(b$mids/1e6,log10(Donor),pch=20,cex=1,col= "#F4E6AA")
  #   maximim<-2
  #   lines(TabSortedCompiled$position/1e6,TabSortedCompiled$Donor*maximim,col="red",lwd=1)
  #   abline(v=galkpos/1e6,lwd=1,col="yellow")
  # }
  # 
  #find the length of segments
  #print("find length")
  Tabtmp<-TabSortedCompiled
  index=which(Tabtmp$Donor==1)
  rec0=1
  
  while(length(index)>0)
  {#print(index[1])
    if(index[1] != 1){
      begin_1 = Tabtmp$position[index[1]]
      begin_0 = Tabtmp$position[index[1] - 1]
    }
    else{
      begin_1 = 0
      begin_0 = 0
    }
    Tabtmp<-Tabtmp[c(index[1]:nrow(Tabtmp)),]
    index=which(Tabtmp$Donor==0)
    if (length(index)>0)
    {end_0=Tabtmp$position[index[1]]
    end_1=Tabtmp$position[index[1]-1]}
    else {
      end_1=4639675
      end_0=4639675}
    FragmentPosions[i,"recombinant"]=j
    
    
    FragmentPosions[i,"rec"]=rec0
    FragmentPosions[i,"begin_1"]=begin_1
    FragmentPosions[i,"begin_0"]=begin_0
    FragmentPosions[i,"end_1"]=end_1
    FragmentPosions[i,"end_0"]=end_0
    ones<-TabSortedCompiled[TabSortedCompiled$position>=begin_1 & TabSortedCompiled$position<end_0,]
    FragmentPosions[i,"nb1"]=sum(ones$source)
    pos<-unique(ones[ones$source==1,]$position)
    #FragmentPosions[i,"nbpos1"]=length(unique(ones[ones$source==1,]$position))
    FragmentPosions[i,"nbpos1"]=length(pos)
    FragmentPosions[i,"length"]=max(pos)-min(pos)
    i<-i+1
    rec0<-rec0+1
    
    if (length(index)==0){break}
    Tabtmp<-Tabtmp[c(index[1]:nrow(Tabtmp)),]
    index=which(Tabtmp$Donor==1)
  }
}

#dev.off()


hist(log10(FragmentPosions$length),breaks=20)
DistREL<-hist(log10(FragmentPosions[FragmentPosions$recombinant<=28,]$length),breaks=seq(0,7,0.25))
DistHS<-hist(log10(FragmentPosions[FragmentPosions$recombinant>28 & FragmentPosions$recombinant<49,]$length),breaks=seq(0,7,0.25))
Dist536<-hist(log10(FragmentPosions[FragmentPosions$recombinant>48,]$length),breaks=seq(0,7,0.25))
```

```{r}
FragmentPositions_filtered <- FragmentPosions[FragmentPosions$length > 600, ]
```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

[Figure 5.C/D/E\
*Code Stackplot*]{.underline}

```{r}
Recs<- FragmentPositions_filtered
Recs$recipient<-"REL606"
Recs$recipient[Recs$recombinant>28 & Recs$recombinant<49]<-"HS"
Recs$recipient[Recs$recombinant>49]<-"536"
for(source in c("REL606", "HS", "536")){
  if (source=="REL606")
  {
    galkpos=769500
    genomeLength=4629812
    color="#155289"
  }
  if (source=="HS")
  {
    galkpos=816000
    genomeLength=4643538
    color="#56B4E9"
  }
  
  if (source=="536")
  {
    galkpos=816000
    genomeLength=(5500000)
    color="pink3"
  }
  nbrecombinants<-length(unique(Recs[Recs$recipient==source,]$recombinant))
  file_name <- paste0("PhD/Figures/Figure5_", source, ".tiff")
tiff(file_name, width = 4.5, height = 4.5, units = "in", res = 500)  # dimensions et résolution

# Initialisation du graphique avec un fond vide
plot(NA, xlim = c(0, genomeLength), ylim = c(0, nbrecombinants), main = source, yaxt = "n", axes = FALSE,
     xlab = "Genome position (Mb)", ylab = "Recombinants", cex.lab = 1.5, cex.axis = 1.5, cex.main = 1.5, cex.sub = 1.5)
axis(side = 1, at = c(0:5) * 1e6, labels = c(0:5))

# Dessiner la ligne rouge pour galkpos
abline(v = galkpos, col = "red", lwd = 3)

# Initialisation de la base pour les polymorphismes
base <- 0
for (i in unique(Recs[Recs$recipient == source,]$recombinant)) {
  tmp <- Recs[Recs$recombinant == i,] 
  pointsD <- data.frame(pos = c(0), rec = c(0))
  pointsD[1,] <- c(0, base + 0.05) 
  
  b <- 1
  for (j in 1:length(tmp[, 1])) {
    pointsD[b + 1,] <- c(tmp[j, "begin_1"], base + 0.05)
    pointsD[b + 2,] <- c(tmp[j, "begin_1"], base + 0.95)
    pointsD[b + 3,] <- c(tmp[j, "begin_1"] + tmp[j, "length"], base + 0.95) 
    pointsD[b + 4,] <- c(tmp[j, "begin_1"] + tmp[j, "length"], base + 0.05)
    b <- b + 4
  }
  pointsD[b + 1,] <- c(genomeLength, base + 0.05)
  pointsD[b + 2,] <- c(0, base + 0.05)
  polygon(c(0, 0, genomeLength, genomeLength), c(base + 0.05, base + 0.95, base + 0.95, base + 0.05), col = color)
  polygon(pointsD, col = "#F4E6AA")
  
  # Mise à jour pour le prochain recombinant
  base <- base + 1
}

dev.off()

}
```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

[Figure 5G]{.underline}

```{r}
couleurs_personnalisees = rev(c("#155289","#679df1", "#56B4E9","#9fe2f2", "pink3","pink" ))

Data_rec_REL <-  subset(Recs, recipient == "REL606")
Data_rec_HS <-  subset(Recs, recipient == "HS")
Data_rec_536 <-  subset(Recs, recipient == "536")
galkpos_REL=769500
galkpos_HS=816000
galkpos_536=803500


#Nombre de fragments recombinés
nb_frag_REL_gal <- nrow(Data_rec_REL)
nb_frag_HS_gal <- nrow(Data_rec_HS)
nb_frag_536_gal <-nrow(Data_rec_536)

#Longueur des fragments recombinés
len_REL_gal <- Data_rec_REL$length
len_HS_gal <- Data_rec_HS$length
len_536_gal <- Data_rec_536$length


REL_galK <- Data_rec_REL[Data_rec_REL$begin_0 <= galkpos_REL & Data_rec_REL$end_0 >= galkpos_REL, ]
len_REL_avec_galK <- REL_galK$length
HS_galK <- Data_rec_HS[Data_rec_HS$begin_0 <= galkpos_HS & Data_rec_HS$end_0 >= galkpos_HS, ]
len_HS_avec_galK <- HS_galK$length
s536_galK <- Data_rec_536[Data_rec_536$begin_0 <= galkpos_536 & Data_rec_536$end_0 >= galkpos_536, ]
len_536_avec_galK <- s536_galK$length


Recipient <- c(rep("Rel606", length(len_REL_gal)), rep("HS", length(len_HS_gal)),rep("536", length(len_536_gal)), rep("Rel606_onlyGal", length(len_REL_avec_galK)), rep("HS_OnlyGal", length(len_HS_avec_galK)),rep("536_OnlyGal", length(len_536_avec_galK)))

Recipient2 <- c(rep("Rel606", length(len_REL_gal)), rep("HS", length(len_HS_gal)),rep("536", length(len_536_gal)))
Recipient_onlyGAL <- c(rep("Rel606_onlyGal", length(len_REL_avec_galK)), rep("HS_OnlyGal", length(len_HS_avec_galK)),rep("536_OnlyGal", length(len_536_avec_galK)))
ALL_length2 <- c(len_REL_gal,len_HS_gal,len_536_gal)
Gal_length <-     c(len_REL_avec_galK,len_HS_avec_galK,len_536_avec_galK)
ALL_length <- c(len_REL_gal,len_HS_gal,len_536_gal,len_REL_avec_galK, len_HS_avec_galK, len_536_avec_galK)

Combined_length <- data.frame(Recipient, ALL_length)
Combined_length_onlyGalK <- data.frame(Recipient_onlyGAL, Gal_length)
Combined_length_sorted <- Combined_length %>% arrange(desc(Recipient))
Total_length <- data.frame(Recipient2, ALL_length2)

Combined_length_sorted <- Combined_length %>% arrange(desc(Recipient))
Combined_length_sorted <- Combined_length_sorted %>%
  mutate(RecipientF = Recipient)

Combined_length_sorted <- Combined_length_sorted %>%
  mutate(new_column = ifelse(grepl("Gal", Recipient), "Gal", "All"))%>%
  mutate(Recipient = str_replace(RecipientF, "Rel606_onlyGal", "Rel606"))%>%
  mutate(Recipient = str_replace(RecipientF, "HS_OnlyGal", "HS"))%>%
  mutate(Recipient = str_replace(RecipientF, "536_OnlyGal", "536"))

tiff("PhD/Figures/Figure5G.tiff", width =5, height = 3, units = "in", res = 500)


ggplot(Combined_length_sorted, aes(x = RecipientF, y = log10(ALL_length), fill = RecipientF)) +
  geom_boxplot(width = 0.5) +  # Boxplot
  geom_jitter(width = 0.1, shape = 21,alpha = 0.8, stroke = 1,col = "black") + 
  labs(y = "Log10(Recombined Length)", x = "Recipient") + 
  scale_fill_manual(values = couleurs_personnalisees) +
  coord_flip() +  
  theme_minimal() +  
  theme(legend.position = "none")+
  geom_rect(aes(xmin = 0.6, xmax = 2.40, ymin = 2, ymax = 6.1), 
            color = "pink3", fill = NA, size = 1, linetype = "solid")+
  geom_rect(aes(xmin = 2.55, xmax = 4.4, ymin = 2, ymax = 6.1), 
            color = "lightblue", fill = NA, size = 1, linetype = "solid")+
  geom_rect(aes(xmin = 4.55, xmax = 6.5, ymin = 2, ymax = 6.1), 
            color = "darkblue", fill = NA, size = 1, linetype = "solid")+
  geom_text(aes(x = 1.7, y = 2.5, label = "536"), 
            color = "pink3", size = 4, family="Arial",fontface = "bold") +
  geom_text(aes(x = 3.7, y = 2.5, label = "HS"), 
            color = "lightblue", size = 4, family="Arial",fontface = "bold") +
  geom_text(aes(x = 5.7, y = 2.5, label = "REL606"), 
            color = "darkblue", size = 4, family="Arial", fontface ="bold")+
  stat_compare_means(
    aes(group = RecipientF), 
    method = "wilcox.test", 
    comparisons = list(c("Rel606_onlyGal", "Rel606"), 
                       c("HS_OnlyGal", "HS"), 
                       c("536_OnlyGal", "536")),
    label = "p.signif", 
    hide.ns = FALSE, 
    size = 3,
    label.x = 1.5,      
    label.y = 6.2,
    tip.length =0
  )+
  stat_compare_means(
    aes(group = RecipientF), 
    method = "wilcox.test", 
    comparisons = list(c("Rel606_onlyGal", "HS_OnlyGal")),
    label = "p.signif", 
    hide.ns = FALSE, 
    size = 3,
    label.x = 1.5,      
    label.y = 1.25,
    tip.length =0
  )+
  stat_compare_means(
    aes(group = RecipientF), 
    method = "wilcox.test", 
    comparisons = list(c("536_OnlyGal", "Rel606_onlyGal")),
    label = "p.signif", 
    hide.ns = FALSE, 
    size = 3,
    label.x = 1.5,      
    label.y = 1,
    tip.length =0
  )+
  stat_compare_means(
    aes(group = RecipientF), 
    method = "wilcox.test", 
    comparisons = list(c("536_OnlyGal", "HS_OnlyGal")),
    label = "p.signif", 
    hide.ns = FALSE, 
    size = 3,
    label.x = 1.5,      
    label.y = 1.5,
    tip.length =0
  )
dev.off()
```

```{r}
mean(FragmentPositions_filtered$length[FragmentPositions_filtered$recombinant < 29])
```

```{r}
mean(FragmentPositions_filtered$length[FragmentPositions_filtered$recombinant < 49 &
                                         FragmentPositions_filtered$recombinant > 28])
```

```{r}
mean(FragmentPositions_filtered$length[FragmentPositions_filtered$recombinant > 48])
```

```{r}
mean(REL_galK$length)
```

```{r}
mean(HS_galK$length)
```

```{r}
mean(s536_galK$length)
```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

[Figure 5F]{.underline}

```{r}

tiff("PhD/Figures/Figure5F.tiff", width = 4.5, height = 4.5, units = "in", res = 500)
Total_length$Recipient2 <- factor(Total_length$Recipient2, levels = c("Rel606", "HS", "536"))
ggplot(Total_length, aes(x = log10(ALL_length2), fill = Recipient2)) +
  geom_histogram(position = "identity", alpha = 0.5, bins = 17, color = "black", size = 0.2) + 
  labs(x = "Log10 (Fragment Length)", y = "Count") + 
  theme_minimal() + 
  scale_fill_manual(values = c("Rel606" = "#155289", "HS" = "#56B4E9", "536" = "pink3")) +  facet_wrap(~Recipient2, scales = "fixed") +  
  theme(legend.position = "none", 
        panel.border = element_rect(color = "black", fill = NA, size = 1)) +  
  coord_fixed(ratio = 1)+
  scale_y_continuous(limits = c(0, 8))
dev.off()
```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

[Figure 5H]{.underline}

```{r}


tiff("PhD/Figures/Figure5H.tiff", width = 4.5, height = 4.5, units = "in", res = 500)
resultat <- Recs 
resultat <- resultat%>%
  group_by(recombinant) %>%
  summarise(Total_Longueur = sum(length))
resultat$Total_Longueur <- as.numeric(resultat$Total_Longueur)
resultat <- resultat %>%
  mutate(recombinant = case_when(
    recombinant >= 9 & recombinant <= 28 ~ "REL606",          # Condition pour 9 à 28
    recombinant > 28 & recombinant <= 48 ~ "HS",              # Condition pour 28 à 48
    recombinant > 48 ~ "536",                                 # Condition pour > 48
    TRUE ~ as.character(recombinant)                          # Cas par défaut pour les autres valeurs
  ))

resultat$recombinant <- factor(resultat$recombinant, levels = c("REL606", "HS", "536"))


ggplot(resultat, aes(x = log10(Total_Longueur), fill = recombinant)) +
  geom_histogram(position = "dodge", alpha = 0.5, bins = 17, color = "black", 
                 linewidth = 0.2) +
  labs(x = "Log10 (Total Length)", y = "Count") + 
  theme_minimal() +
  scale_fill_manual(
    values = c("REL606" = "#155289", "HS" = "#56B4E9", "536" = "pink3"),
    labels = c("REL606", "HS", "536"), 
    name = "Recipient" 
  ) +  facet_wrap(~recombinant, scales = "fixed")+
  theme(
    legend.position = "none", 
    panel.border = element_rect(color = "black", fill = NA, linewidth = 1))+
  coord_fixed(ratio = 1)  +
  scale_x_continuous(limits = c(3, 6.1))+
  scale_y_continuous(limits = c(0, 8))
dev.off()
```

------------------------------------------------------------------------

```{r}
RecSize_onlyGal <- RecSize[RecSize$begin_0 <= galkpos_REL & RecSize$end_0 >= galkpos_REL, ]
Combined_length_onlyGalK$Recipient_onlyGAL <- factor(Combined_length_onlyGalK$Recipient_onlyGAL,levels = c("Rel606_onlyGal", "HS_OnlyGal", "536_OnlyGal"))
```

Save the file with the recombined fragment data

```{r}
Table_pos <- FragmentPositions_filtered 
Table_pos<- Table_pos[, -c(3,4,6)] 
write.csv(Table_pos, file = "Table_pos_OT1601.csv", row.names = FALSE)

Table_pos_unfiltered <- FragmentPosions
Table_pos_unfiltered<- Table_pos_unfiltered[, -c(3,4,6)] 
write.csv(Table_pos, file = "Table_pos_unfilt_OT1601.csv", row.names = FALSE)
```

------------------------------------------------------------------------

[Figure 6A/B/C.]{.underline}

Pool analysis

```{r}
galkpos_REL=769500
genomeLength_REL=4629812
galkpos_HS=816000
genomeLength_HS=4643538
galkpos_536=816000
genomeLength_536=5500000


recipient_colors <- c("#155289", "#56B4E9", "pink3")  
i <- 1
par(mfrow = c(2,2))
j_vec = c(LETTERS[1:3])
terpos = 1500000
ratio = list()
list_max_ratio = list()
list_max = list()
store = list()
store3 = list() 
genomeLength = 4629812
#dev.new(width = 8, height = 6)  
for (v in 1:length(j_vec)) {
  j = j_vec[v]
  print(v)
  
  Donor <- seq(0, 5000000, 1000) * 0
  Recipient <- seq(0, 5000000, 1000) * 0
  Donor <- Donor[-1]
  Recipient <- Recipient[-1]
  
  # A= K12 x Rel B= K12 x HS and C = K12 x 536
  name2 = paste("data/GalK/Pools/", j, "pool_K12.sam", sep="")
  name1 = paste("data/GalK/Pools/", j, "pool_Rec.sam", sep="")
  print(name2)
  Tab <- read.table(name1)
  Tab2 <- read.table(name2)
  
  a <- hist(Tab[Tab[, 2] >= 0, 1], breaks = seq(0, 5000000, 1000), plot = F)
  b <- hist(Tab2[Tab2[, 2] >= 0, 2], breaks = seq(0, 5000000, 1000), plot = F)
  Donor <- Donor + b$counts
  Recipient <- Recipient + a$counts
  
  # Combine all reads according to the recipient position and attribute to who they matched, keeping only reads that matched both genomes but preferentially one
  TabSorted1 <- as.data.frame(Tab[Tab[, 2] >= 0 & abs(Tab[, 1] - Tab[, 2]) < 500000, 1])
  TabSorted1$source <- 0
  # 0 means recipient match
  dimnames(TabSorted1)[[2]][1] <- "position"
  
  TabSorted2 <- as.data.frame(Tab2[Tab2[, 2] >= 0 & abs(Tab2[, 1] - Tab2[, 2]) < 500000, 2])
  TabSorted2$source <- 1
  # 1 means donor match
  dimnames(TabSorted2)[[2]][1] <- "position"
  TabSortedCompiled <- rbind(TabSorted1, TabSorted2)
  
  # Combine the reads info of both tables, 0 is for reads that match the recipient, 1 for the ones that match the donor  
  TabSortedCompiled <- TabSortedCompiled[order(TabSortedCompiled$position), ]
  
  combined <- cbind(as.data.frame(Donor), Recipient)
  combined <- cbind(as.data.frame(combined), a$mids)
  names(combined)[3] <- "xaxis"
  
  max = combined$xaxis[which.max(combined$Donor)]
  list_max = append(list_max, max)  
  
  max_ratio = combined$xaxis[which.max(Donor / Recipient)]
  list_max_ratio = append(list_max_ratio, max_ratio)
  
  # Renommé 'list' en 'ratio_list'
  ratio_list = log10(Donor / Recipient)
  ratio[[j]] = ratio_list
  
  

  if(j=="A"){
    #tiff("PhD/Figure6A.tiff", width = 4.5, height = 4.5, units = "in", res = 500)
    p <- ggplot(combined, aes(x = xaxis / 1e6, y = log10(Donor))) +
      geom_vline(xintercept = 
                   ifelse(v == 1, galkpos_REL / 1e6, 
                          ifelse(v == 2, galkpos_HS / 1e6, galkpos_536 / 1e6)), 
                 color = "red", linewidth = 1, linetype = "solid") +  
      geom_point(aes(y = log10(Recipient)), fill = recipient_colors[1], color = "black", stroke = 0.05,
                 shape = 21, size = 2) +  # Points pour Recipient
      geom_point(aes(y = log10(Donor)), fill = "#F4E6AA", shape = 21, size = 2, color = "black", stroke = 0.05) +  
      labs(x = "Genome position (Mb)") +
      scale_y_continuous(limits = c(0, 3)) +
      theme(
        panel.background = element_rect(fill = "white"),  # Fond blanc
        panel.border = element_rect(colour = "black", fill = NA, size = 1),  # Bordure noire autour du plot
        plot.background = element_rect(fill = "white"),  # Fond général blanc
        axis.title = element_text(size = 14),  # Taille du titre des axes
        axis.text = element_text(size = 12),  # Taille des labels des axes
        plot.title = element_text(size = 16, hjust = 0.5),  # Titre centré et taille
        plot.subtitle = element_text(size = 14, hjust = 0.5))  # Sous-titre centre
    ggsave("PhD/Figures/Figure6A.tiff", 
           plot = p,  
           width = 4.5, 
           height = 4.5, 
           units = "in", 
           dpi = 500,  # Résolution de l'image
           device = "tiff")  # Format tiff
    
    DR <- log10(Donor / Recipient)
    store3 <- cbind(store3, DR[c((floor(genomeLength_REL / 1000) - (1500 - floor(galkpos_REL / 1000))):
                                       (floor(genomeLength_REL / 1000)), 1:floor(galkpos_REL / 1000 + 1500))])
  }
  if(j=="B"){
    # Assure-toi que ton graphique ggplot est correctement généré
    p <- ggplot(combined, aes(x = xaxis / 1e6, y = log10(Donor))) +
      geom_vline(xintercept = 
                   ifelse(v == 1, galkpos_REL / 1e6, 
                          ifelse(v == 2, galkpos_HS / 1e6, galkpos_536 / 1e6)), 
                 color = "red", linewidth = 1, linetype = "solid") +  
      geom_point(aes(y = log10(Recipient)), fill = recipient_colors[2], color = "black", stroke = 0.05,
                 shape = 21, size = 2) +  # Points pour Recipient
      geom_point(aes(y = log10(Donor)), fill = "#F4E6AA", shape = 21, size = 2, color = "black", stroke = 0.05) +  
      labs(x = "Genome position (Mb)") +
      scale_y_continuous(limits = c(0, 3)) +
      theme(
        panel.background = element_rect(fill = "white"),  # Fond blanc
        panel.border = element_rect(colour = "black", fill = NA, size = 1),  
        plot.background = element_rect(fill = "white"),  
        axis.title = element_text(size = 14),  
        axis.text = element_text(size = 12), 
        plot.title = element_text(size = 16, hjust = 0.5),  
        plot.subtitle = element_text(size = 14, hjust = 0.5)) 
    
    p <- ggsave("PhD/Figures/Figure6B.tiff", 
           plot = p,  # Plot à enregistrer
           width = 4.5, 
           height = 4.5, 
           units = "in", 
           dpi = 500, 
           device = "tiff") 
    
        DR <- log10(Donor / Recipient)
    store3 <- cbind(store3, DR[c((floor(genomeLength_HS / 1000) - 
                                        (1500 - floor(galkpos_HS / 1000))):
                                       (floor(genomeLength_HS / 1000)), 
                                     1:floor(galkpos_HS / 1000 + 1500))])}
  if(j=="C"){
    p<- ggplot(combined, aes(x = xaxis / 1e6, y = log10(Donor))) +
      geom_vline(xintercept = 
                   ifelse(v == 1, galkpos_REL / 1e6, 
                          ifelse(v == 2, galkpos_HS / 1e6, galkpos_536 / 1e6)), 
                 color = "red", linewidth = 1, linetype = "solid") +  
      geom_point(aes(y = log10(Recipient)), fill = recipient_colors[3], color = "black", stroke = 0.05,
                 shape = 21, size = 2) +  # Points pour Recipient
      geom_point(aes(y = log10(Donor)), fill = "#F4E6AA", shape = 21, size = 2, color = "black", stroke = 0.05) +  
      labs(x = "Genome position (Mb)") +
      scale_y_continuous(limits = c(0, 3)) +
      theme(
        panel.background = element_rect(fill = "white"),  # Fond blanc
        panel.border = element_rect(colour = "black", fill = NA, size = 1), 
        plot.background = element_rect(fill = "white"),
        axis.title = element_text(size = 14),  
        axis.text = element_text(size = 12), 
        plot.title = element_text(size = 16, hjust = 0.5), 
        plot.subtitle = element_text(size = 14, hjust = 0.5))  
    ggsave("PhD/Figures/Figure6C.tiff", 
           plot = p, 
           width = 4.5, 
           height = 4.5, 
           units = "in", 
           dpi = 500,  
           device = "tiff")  
    DR <- log10(Donor / Recipient)
    store3 <- cbind(store3, DR[c((floor(genomeLength_536 / 1000) - (1500 - floor(galkpos_536 / 1000))):
                                       (floor(genomeLength_536 / 1000)), 1:floor(galkpos_536 / 1000 + 1500))])
  }
}

```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

[Figure 6.D.]{.underline}

```{r echo=TRUE, warning=FALSE, message=FALSE}

### Position relative to galK 
tiff("PhD/Figures/Figure6D.tiff", width = 4.5, height = 4.5, units = "in", res = 500)

plot(NA,xlim=c(-1500000,1500000)/1000000,ylim=c(-3.5,3),ylab="log10(Donor/Recipient)"
           ,xlab="Genome position relative to the GalK",cex.lab=1, 
           cex.axis=1, cex.main=1, cex.sub=1.5)
abline(v = 0, col = "red", lwd = 3)  # Rouge et épaisseur 2
points(seq(-1502000,1498000,1000)/1000000,store3[,1],pch=20,col="#155289",cex=0.5)
points(seq(-1502000,1498000,1000)/1000000 ,store3[,2],pch=20,col="#56B4E9",cex=0.5)
points(seq(-1502000,1498000,1000)/1000000,store3[,3],pch=20,col="pink3",cex=0.5)
dev.off()
```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

[Figure 6.E]{.underline}

```{r echo=TRUE, warning=FALSE, message=FALSE}
ratio <- as.data.frame(ratio)
ratio <- cbind(ratio, combined$xaxis)
names(ratio)[4] <- "xaxis"
ratio$xaxis <- as.numeric(ratio$xaxis)
ratio <- ratio[, -c(5:9)]


a <- c(1, Inf, NA, NaN)
ratio_dup <- as.data.frame(ratio)
ratio_dup[ratio_dup == -Inf] <- 0
ratio_dup[ratio_dup == Inf] <- 0
tiff("PhD/Figures/Figure6E.tiff", width = 4.5, height = 4.5, units = "in", res = 500)
print(plot(NA,xlim=c(0,5000000/1000000),ylim=c(-3.5,3),ylab="log10(Donor/Recipient)"
           ,xlab="Genome position (Mb)",cex.lab=1, 
           cex.axis=1, cex.main=1, cex.sub=1.5))
rect(1.2, -3.5, 2, 3, col = rgb(0.4, 0.6, 1, 0.3), border = NA)  # Rectangle bleu clair
rect(3.7, -3.5, 4.5, 3, col = rgb(0.2, 0.8, 0.2, 0.3), border = NA)  # Rectangle vert
abline(v = galkpos_REL / 1000000, col = "red", lwd = 3)  
points(ratio$xaxis/1000000, ratio$A,pch=20,col="#155289",cex=0.5)
text(4.1, 3, "Ori", cex = 1, col = "darkgreen", font = 4)
text(1.6, 3, "Ter", cex = 1, col = "darkblue", font = 4)
dev.off()
```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

[Figure 6.F]{.underline}

```{r echo=TRUE, warning=FALSE, message=FALSE}

i<-1
k_vec = c(1:5)
par(mfrow = c(2,2))
j_vec = c(LETTERS[1:5],LETTERS[8:14])
p_vec = c("", "pool")
galkpos=769500
terpos = 1500000
ratio = list()
list_max_ratio = list()
list_max = list()
store= list()
sotre = list()
genomeLength = 4629812

for (v in 1:length(j_vec)) {
  j = j_vec[v]
  k=k_vec[v]
  
  Donor<-seq(0,5000000,1000)*0
  Recipient<-seq(0,5000000,1000)*0
  Donor<-Donor[-1]
  Recipient<-Recipient[-1]
  
  # name2=paste("position/K12_", j, "1pool_pos.txt",sep="")
  #  name1=paste("position/Rel606_", j,"1pool_pos.txt",sep="")
  name2=paste("data/POOLS/", j, "1pool_K12.sam",sep="")
  name1=paste("data/POOLS/", j, "1pool_Rel606.sam",sep="")
  print(name2)
  Tab<-read.table(name1)
  Tab2<-read.table(name2)
  
  a<-hist(Tab[Tab[,2]>=0,1],breaks=seq(0,5000000,1000),plot=F)
  b<-hist(Tab2[Tab2[,2]>=0,2],breaks=seq(0,5000000,1000),plot=F)
  Donor<-Donor+b$counts
  Recipient<-Recipient+a$counts
  
  #combine all reads according to the recipient position and attibute to who they matched, keeping only reads that matched both genomes but preferentially one
  TabSorted1<-as.data.frame(Tab[Tab[,2]>=0 & abs(Tab[,1]-Tab[,2])<500000,1])
  TabSorted1$source<-0
  #0 means recipient match
  dimnames(TabSorted1)[[2]][1]<-"position"
  
  TabSorted2<-as.data.frame(Tab2[Tab2[,2]>=0 & abs(Tab2[,1]-Tab2[,2])<500000,2])
  TabSorted2$source<-1
  #1 means donor match
  dimnames(TabSorted2)[[2]][1]<-"position"
  TabSortedCompiled<-rbind(TabSorted1,TabSorted2)
  
  #combine the reads info of both table 0 is for reads that match the recipient, 1 for the ones that match the donor  
  TabSortedCompiled<-TabSortedCompiled[order(TabSortedCompiled$position),]
  
  
  
  combined = cbind(as.data.frame(Donor), Recipient)
  combined = cbind(as.data.frame(combined), a$mids)
  names(combined)[3] <- "xaxis"
  max = combined$xaxis[which.max(combined$Donor)]
  list_max = append(list_max, max)  
  
  max_ratio=combined$xaxis[which.max(Donor/Recipient)]
  list_max_ratio = append(list_max_ratio, max_ratio)
  
  
  # 
  # print(ggplot(combined, aes(x=xaxis/1e6, y=log10(Donor)))+
  #         geom_point(aes(y = log10(Recipient)), col = "#155289")+
  #         geom_point(aes(y = log10(Donor)), col = "#F4E6AA")+
  #         labs(title = paste("Recombinant", j, "pool"),
  #              subtitle = paste("insertion at:", max), x = "genome (Mb)") +
  #         geom_vline(xintercept = max/1e6))
  
  # maximim<-2
  # 
  # a <- plot(a$mids/1e6,log10(Recipient),pch=20,cex=1,main=paste("Recombinant",j), 
  #      xlab="genome position (Mb)",ylab="log10(coverage)",ylim=c(0,min(3.5,max(log10(Recipient),log10(Donor)))),
  #      col="blue",cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5)
  # points(b$mids/1e6,log10(Donor),pch=20,col= "#F4E6AA",cex=1)
  # lines(TabSortedCompiled$position/1e6,TabSortedCompiled$Donor*maximim,col="red",lwd=1)
  # abline(v=max/1e6,lwd=1,col="yellow")
  # 
  
  
  list = log10(Donor/Recipient)
  ratio[[j]] = list
  
  DR<-log10(Donor/Recipient)
  store<-cbind(store,DR[c((floor(genomeLength/1000)-(1500-floor(terpos/1000))):
                            (floor(genomeLength/1000)),1:floor(terpos/1000+1500))])
  DR<-log10(Donor/Recipient)
  store3<-cbind(store,DR[c((floor(genomeLength/1000)-(1500-floor(galkpos/1000))):
                            (floor(genomeLength/1000)),1:floor(galkpos/1000+1500))])

}

dev.off()


ratio = as.data.frame(ratio)
ratio = cbind(ratio, combined$xaxis)
names(ratio)[13] <- "xaxis"
a <- c(1, Inf, NA, NaN)
ratio_dup <- as.data.frame(ratio)
ratio_dup[ratio_dup == -Inf] <- 0
ratio_dup[ratio_dup == Inf] <- 0

tiff("PhD/Figures/Figure6F.tiff", width = 4.5, height = 4.5, units = "in", res = 500)
print(plot(NA,xlim=c(0,5000000/1000000),ylim=c(-3,3),ylab="log10(Donor/Recipient)"
           ,xlab="Genome position (Mb)",cex.lab=1, 
           cex.axis=1, cex.main=1, cex.sub=1.5))
rect(1.2, -3.5, 2, 3, col = rgb(0.4, 0.6, 1, 0.3), border = NA)  # Rectangle bleu clair
rect(3.7, -3.5, 4.5, 3, col = rgb(0.2, 0.8, 0.2, 0.3), border = NA)  # Rectangle vert
abline(v = galkpos_REL / 1000000, col = "red", lwd=3)
points(ratio$xaxis/1000000, ratio$K,pch=20,col="darkblue",cex=0.5)
points(ratio$xaxis/1000000, ratio$I,pch=20,col="#56B4E9",cex=0.5)
points(ratio$xaxis/1000000, ratio$H,pch=20,col="lightblue",cex=0.5)
points(ratio$xaxis/1000000, ratio$J,pch=20,col="purple3",cex=0.5)
text(4.1, 3, "Ori", cex = 1, col = "darkgreen", font = 4)
text(1.6, 3, "Ter", cex = 1, col = "darkblue", font = 4)
dev.off()

```

------------------------------------------------------------------------

[Figure 7]{.underline}

```{r echo=TRUE, warning=FALSE, message=FALSE}
tiff("PhD/Figures/Figure7.tiff", width = 4.5, height = 4.5, units = "in", res = 500)
par(mar =c(5, 4, 4, 6), pty = "s")

print(plot(NA,xlim=c(0,5000000/1000000),ylim=c(-3.5,3),ylab="log10(Donor/Recipient)"
           ,xlab="Genome position (Mb)",cex.lab=1, 
           cex.axis=1, cex.main=1, cex.sub=1.5))
abline(v = which.min(ratio_dup$N)/1000, col = "black", lwd = 1, lty = 2)  
points(ratio$xaxis/1000000, ratio$L,pch=20,col="brown",cex=0.5)
points(ratio$xaxis/1000000, ratio$M,pch=20,col="gold3",cex=0.5)
points(ratio$xaxis/1000000, ratio$N,pch=20,col="darkgreen",cex=0.5)
abline(v = ratio_dup$xaxis[which.max(ratio_dup$L)]/1000000, col = "darkred")
abline(v = ratio_dup$xaxis[which.max(ratio_dup$M)]/1000000, col = "gold3")
abline(v = ratio_dup$xaxis[which.max(ratio_dup$N)]/1000000, col = "darkgreen")
dev.off()
```

------------------------------------------------------------------------

------------------------------------------------------------------------

------------------------------------------------------------------------

[Supplementary Figures\
]{.underline}The data have been processed with 1) BWA then 2) with .perl
file to obtain the final .sam

```{r echo=TRUE, warning=FALSE, message=FALSE}

tab<-read.table("data/Tnseq/data_paper/K12_TnPos.sam")
#filter ther reads 
restricted<-tab[tab$V5>(-15)&tab$V5<(-8), ]
dimnames(restricted)[[2]]=c("read","UMI","pos","strand","offset")
resun<-unique(restricted[c("UMI","pos"),])



restricted$dup <- duplicated(restricted[c("UMI", "pos") ])
dedourrestricted<- subset(restricted, dup == FALSE)

counts <-hist(dedourrestricted$pos,breaks=seq(0,4639675+250000,250000))
counts1 <-hist(unique(dedourrestricted$pos),breaks=seq(0,4639675+250000,250000),plot=FALSE)
boby0=mean(counts$counts[6:7])
boby1=mean(counts1$counts[6:7])
```

```{r}
# Premier graphique : log2 fold increase coverage
tiff("PhD/Figures/Supplementary_FiguresTn1.tiff", width = 4.5, height = 4.5, units = "in", res = 500) 
plot(counts$mids / 1000000, log2(counts$counts / boby0), xlab = "Genome Position (Mb)", pch = 20, col = "blue", 
     ylim = c(-1, 5), cex = 2, ylab = "log2 fold increase compared to terminus")
points(counts1$mids / 1000000, log2(counts1$counts / boby1), pch = 20, col = "red")
abline(h = 0, lty = "dotted")
dev.off()
```

```{r}

# Deuxième graphique : graphique de régression entre les deux log2 fold increases
tiff("PhD/Figures/Supplementary_FiguresTn2.tiff", width = 4.5, height = 4.5, units = "in", res = 500) 
plot(log2(counts1$counts / boby1), log2(counts$counts / boby0), xlab = "relative log2 fold increase\n in unique pos", 
     ylab = "relative log2 fold increase in library insertions", pch = 20, cex = 2, col = "violet")
a <- lm(log2(counts$counts / boby0) ~ log2(counts1$counts / boby1))
abline(a = a$coefficients[1], b = a$coefficients[2], lty = "dotted", col = "violet")
dev.off()
```
