## ğŸ‘©â€ğŸ’» Codes

This folder contains the following files:  

ğŸ“‚ codes/  
â”œâ”€â”€ ğŸ“„ PaperConj2025_analysis_090925.Rmd  
â”œâ”€â”€ ğŸ“„ bwa_25.py  
â”œâ”€â”€ ğŸ“„ analysis_downscaling.py  
â”œâ”€â”€ ğŸ“„ downscaling.sh  
â”œâ”€â”€ ğŸ“„ ecoli_25.py  
â”œâ”€â”€ ğŸ“„ CreateFilesForHaplotypingSamples_Recombination.pl  
â”œâ”€â”€ ğŸ“„ HaplotypingSampleRecombination.pl  
â”œâ”€â”€ ğŸ“„ automatizationStep2.py  
â””â”€â”€ ğŸ“„ SortReads.pl  

____

**R files**
- `PaperConj2025_analysis_090925.Rmd`    
<small>This .R notebook regroup all the codes needed for the paper main and supplementary figures. The file uses different inputs that are either the supplementary tables or output files (e;g. .txt, .sam..) of previous perl/python pipeline.</small>


**Bash files**
- `downscaling.sh`    
<small>Using seqtk lower the sequencing resolution of chosen fastq files. </small>


**python files**
- `bwa_25.py`
<small> Automatize bwa. Old pipeline</small>
- 'analysis_downscaling.py' calculated the coverage of fastq files
- `ecoli_25.py`  
<small> These python files are used for .fastaq aligned by BWA using `bwa_25.py` and their parsing automatized in a `ecoli_25.py` Old pipeline </small>
- ` automatizationStep2.py`
- <small> automatization of ` HaplotypingSampleRecombination.pl `</small>

**perl files**  
- `SortReads.pl`  
<small> This .pl is an alternative to `ecoli_25.py` after BWA aligment. Old pipeline </small>
-  `CreateFilesForHaplotypingSamples_Recombination.pl  `
<small>new pipeline for recombinant analysis Step 1 (using parsnp.xmfa as an input : donor against recipient)</small>
- ` HaplotypingSampleRecombination.pl `
<small> new pipeline for recombinant analysis Step 2</small>
